### PR TITLE
fix(gatsby-image): add matchMedia to fix wrong aspect ratio and dimensions in art-directed image arrays

### PR DESCRIPTION
**Pull request by timhagn**
## Description

This PR introduces a `matchMedia()` query for `fluid` or `fixed` art-directed arrays.
This prevents retrieving the wrong image (the first from the art-direction stack ain't always the correct one) when ascertaining its `aspectRatio` or `width & height` respectively.

It adds tests for `fluid` or `fixed` art-directed stacks as well.

_Of course this sadly only works during rendering in the browser_, and falls back to the default first image from the array, but at least in the browser the correct image is chosen ; ).

## Related Issues

Addresses #15189
Addresses a few of the pain points in ##13395
(And probably a few other issues ; )

Best,

Tim.

P.S.: Works quite well for me in `gatsby-backrgound-image` ; ).

### CODE REVIEW

File: packages/gatsby-image/src/__tests__/index.js [hunk 1]
```diff
@@ -119,6 +119,22 @@ const setupImages = (
 }
 
 describe(`<Image />`, () => {
+  const OLD_MATCH_MEDIA = window.matchMedia
```
Comment (pvdz) [inline packages/gatsby-image/src/__tests__/index.js]: These tests run in nodejs which has no `window`. How are we going to test them there?

File: packages/gatsby-image/src/index.js [hunk 2]
```diff
@@ -48,18 +48,67 @@ const convertProps = props => {
   return convertedProps
 }
 
+/**
+ * Checks if fluid or fixed are art-direction arrays.
+ * @param props   {object}   The props to check for images.
+ * @param prop    {string}   Check for fluid or fixed.
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: Please update the jsdoc with proper types. `prop` is an enum and `props` can be copied from `getImageSrcKey` below.

Example for enum:

```js
/**
 * @param prop {"fluid" | "fixed"}
 */
```

File: packages/gatsby-image/src/index.js [hunk 3]
```diff
@@ -48,18 +48,67 @@ const convertProps = props => {
...
+ * @return {boolean}
+ */
+const hasArtDirectionSupport = (props, prop) =>
+  props[prop] &&
+  Array.isArray(props[prop]) &&
+  props[prop].some(image => typeof image.media !== `undefined`)
+
+/**
+ * Checks for fluid or fixed Art direction support.
+ * @param props   {object}   The props to check for art-direction support.
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: Same as above, please update jsdocs

File: packages/gatsby-image/src/index.js [hunk 4]
```diff
@@ -48,18 +48,67 @@ const convertProps = props => {
...
+ * @return {boolean}
+ */
+const hasArtDirectionArray = props =>
+  hasArtDirectionSupport(props, `fluid`) ||
+  hasArtDirectionSupport(props, `fixed`)
+
+/**
+ * Tries to detect if a media query matches the current viewport.
+ * @param media   {string}  A media query string.
+ * @return {*|boolean}
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: I think we should drop the `*` here ("any") and prefix the `window.matchMedia` with a double bang (`!!window.matchMedia(...)`) to ensure it's always a boolean.

File: packages/gatsby-image/src/index.js [hunk 5]
```diff
@@ -48,18 +48,67 @@ const convertProps = props => {
...
-  const data = (fluid && fluid[0]) || (fixed && fixed[0])
+  const data = getCurrentSrcData({ fluid, fixed })
 
   return data.src
 }
 
+/**
+ * Returns the current src - if possible with art-direction support.
+ * @param fluid         {object}    Fluid Image (Array) if existent.
+ * @param fixed         {object}    Fixed Image (Array) if existent.
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: Please add the object typing here. You can copy it from the `getImageSrcKey` jsdoc.

Additionally, note that these are properties, not params. However, as suggested below, the param would be one object that is either the fixed or fluid. You can copy the type from the `getImageSrcKey` jsdoc.

File: packages/gatsby-image/src/index.js [hunk 6]
```diff
@@ -48,18 +48,67 @@ const convertProps = props => {
...
+ * @param index         {number}    The index of the image to return (default: 0).
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: This param seems to be unused (none of the callers use it). Would suggest to make it a constant.

File: packages/gatsby-image/src/index.js [hunk 7]
```diff
@@ -48,18 +48,67 @@ const convertProps = props => {
...
+  props[prop].some(image => typeof image.media !== `undefined`)
+
+/**
+ * Checks for fluid or fixed Art direction support.
+ * @param props   {object}   The props to check for art-direction support.
+ * @return {boolean}
+ */
+const hasArtDirectionArray = props =>
+  hasArtDirectionSupport(props, `fluid`) ||
+  hasArtDirectionSupport(props, `fixed`)
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: Can you please do `hasArtDirectionSupport(props.fluid) || hasArtDirectionSupport(props.fixed)` here? Or is there a reason for passing on the property name separately?

File: packages/gatsby-image/src/index.js [hunk 8]
```diff
@@ -48,18 +48,67 @@ const convertProps = props => {
...
 }
 
+/**
+ * Returns the current src - if possible with art-direction support.
+ * @param fluid         {object}    Fluid Image (Array) if existent.
+ * @param fixed         {object}    Fixed Image (Array) if existent.
+ * @param index         {number}    The index of the image to return (default: 0).
+ * @return {*}
+ */
+const getCurrentSrcData = ({ fluid, fixed }, index = 0) => {
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: As suggested below, please change the parameters here to accept one array (which is either the fluid or fixed, but that distinction doesn't appear to be relevant to tihs function so there's no point in tracking it).
```suggestion
const getCurrentSrcData = currentData => {
```

File: packages/gatsby-image/src/index.js [hunk 9]
```diff
@@ -48,18 +48,67 @@ const convertProps = props => {
...
  */
 const getImageSrcKey = ({ fluid, fixed }) => {
-  const data = (fluid && fluid[0]) || (fixed && fixed[0])
+  const data = getCurrentSrcData({ fluid, fixed })
 
   return data.src
 }
 
+/**
+ * Returns the current src - if possible with art-direction support.
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]:
```suggestion
 * Returns the current src - Preferably with art-direction support.
```

File: packages/gatsby-image/src/index.js [hunk 10]
```diff
@@ -48,18 +48,67 @@ const convertProps = props => {
...
+/**
+ * Checks if fluid or fixed are art-direction arrays.
+ * @param props   {object}   The props to check for images.
+ * @param prop    {string}   Check for fluid or fixed.
+ * @return {boolean}
+ */
+const hasArtDirectionSupport = (props, prop) =>
+  props[prop] &&
+  Array.isArray(props[prop]) &&
+  props[prop].some(image => typeof image.media !== `undefined`)
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: Just curious; is `.media` an "art direction"?

File: packages/gatsby-image/src/index.js [hunk 11]
```diff
@@ -416,7 +465,7 @@ class Image extends React.Component {
 
     if (fluid) {
       const imageVariants = fluid
-      const image = imageVariants[0]
+      const image = getCurrentSrcData({ fluid })
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: I would just pass on `fluid` as an argument, an object seems overkill here.

File: packages/gatsby-image/src/index.js [hunk 12]
```diff
@@ -516,7 +565,7 @@ class Image extends React.Component {
 
     if (fixed) {
       const imageVariants = fixed
-      const image = imageVariants[0]
+      const image = getCurrentSrcData({ fixed })
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: Like the other call, pass on `fixed` as an argument. They have the same structure anyways.

File: packages/gatsby-image/src/index.js [hunk 13]
```diff
@@ -48,18 +48,67 @@ const convertProps = props => {
...
+
 /**
  * Find the source of an image to use as a key in the image cache.
  * Use `the first image in either `fixed` or `fluid`
  * @param {{fluid: {src: string}[], fixed: {src: string}[]}} args
  * @return {string}
  */
 const getImageSrcKey = ({ fluid, fixed }) => {
-  const data = (fluid && fluid[0]) || (fixed && fixed[0])
+  const data = getCurrentSrcData({ fluid, fixed })
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: With other sugestions, this would become `fluid ? getImageSrcKey(fluid) : getImageSrcKey(fixed)`

File: packages/gatsby-image/src/__tests__/index.js (hunk 1 above)
Comment (timhagn) [reply packages/gatsby-image/src/__tests__/index.js]: The `devDependencies` lists `jest-environment-jsdom-fourteen`, so we have access to the window object through `js-dom` in the tests.

File: packages/gatsby-image/src/index.js (hunk 10 above)
Comment (timhagn) [reply packages/gatsby-image/src/index.js]: As the `media` property has to be added by a dev to use art-direction support, I guess we might use them synonymously ; ).

File: packages/gatsby-image/src/index.js (hunk 11 above)
Comment (timhagn) [reply packages/gatsby-image/src/index.js]: The reasoning behind using them as an object was just being able to throw in `props` if needed...

**Comment by timhagn** [comment]
Hi @pvdz,

thanks for the reviews, gonna continue attending them now. I just took the `getCurrentSrcData()` from `gatsby-background-image` and there is quite some use for just using `fluid | fixed` as an object (e.g. directly being able to pass `(converted)props` into the functions), but I see your point for `gatsby-image`. Am in the code : ).

Best,

Tim.

File: packages/gatsby-image/src/index.js [hunk 14]
```diff
@@ -48,18 +48,54 @@ const convertProps = props => {
...
-  const data = (fluid && fluid[0]) || (fixed && fixed[0])
+  const data = fluid ? getCurrentSrcData(fluid) : getCurrentSrcData(fixed)
 
   return data.src
 }
 
+/**
+ * Returns the current src - Preferably with art-direction support.
+ * @param currentData  {array}   The fluid or fixed image array.
+ * @return {*}
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: This PR changes the objects to have an optional `media` property so please reflect this in hte JSdocs.

```
 * @param currentData  {{src: string, media?: string}[]}   The fluid or fixed image array.
```

Since it returns an element of this array in all branches, you can replace the star with that type (`{src: string, media?: string}`), too.

(This goes for all the jsdocs here :) )

File: packages/gatsby-image/src/index.js [hunk 15]
```diff
@@ -48,18 +48,54 @@ const convertProps = props => {
...
+ * @param media   {string}  A media query string.
+ * @return {boolean}
+ */
+const matchesMedia = ({ media }) =>
+  media ? isBrowser && !!window.matchMedia(media).matches : false
+
 /**
  * Find the source of an image to use as a key in the image cache.
  * Use `the first image in either `fixed` or `fluid`
  * @param {{fluid: {src: string}[], fixed: {src: string}[]}} args
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: Please update this jsdocs as well (add the optional `media` field)

File: packages/gatsby-image/src/index.js [hunk 16]
```diff
@@ -48,18 +48,54 @@ const convertProps = props => {
   return convertedProps
 }
 
+/**
+ * Checks if fluid or fixed are art-direction arrays.
+ *
+ * @param currentData  {array}   The props to check for images.
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: The `array` and `object` types don't add much value in terms of typing. They might as well be `any` or `*` here. So please be specific in what the input type is here (see below)

File: packages/gatsby-image/src/index.js [hunk 17]
```diff
@@ -48,18 +48,54 @@ const convertProps = props => {
...
+ * @return {boolean}
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: Nit: the output is only a bool if `currentData` is not falsy (like `undefined`). It's probably best to use `!!currentData` in line 59

File: packages/gatsby-image/src/index.js [hunk 18]
```diff
@@ -48,18 +48,54 @@ const convertProps = props => {
...
+ */
+const hasArtDirectionSupport = currentData =>
+  currentData &&
+  Array.isArray(currentData) &&
+  currentData.some(image => typeof image.media !== `undefined`)
+
+/**
+ * Tries to detect if a media query matches the current viewport.
+ * @param media   {string}  A media query string.
```
Comment (pvdz) [inline packages/gatsby-image/src/index.js]: This is a (destructured) `@property`, not a `@param`

**Comment by timhagn** [comment]
Hi @pvdz,

thanks again for the reviews and sorry that I didn't get back to you earlier, was just fully booked last week and got me a stomach bug over the weekend : (.
Gonna try getting your change requests in this week, though.

Best,

Tim.

**Comment by timhagn** [comment]
Hi @pvdz,

while having a short break of both my bug & work and looking through your requests, I finally realize, why most of `gatsby-image` doesn't use JSDOC at all ^^.
But am at it.

Best, Tim.

**Comment by timhagn** [comment]
Hi again, hope I now id change all comments and requests accordingly : ).

Best,

Tim.

**Comment by timhagn** [comment]
Mkay. The starters_validate failing seems to have nothing to do with this PR, any ideas?

Best,

Tim.

File: packages/gatsby-image/src/index.js (hunk 17 above)
Comment (timhagn) [reply packages/gatsby-image/src/index.js]: Added : ).

**Comment by timhagn** [comment]
Thanks @pvdz!

Hope I didn't get the JSDocs too wrong ^^. Glad something from `gbi` made it back to `gatbsy-image` : ).

Best,

Tiim.
